# 📊 Notebook 4: Data Collection
**Project:** TalentSight — Employee Attrition Intelligence Platform  
**Author:** Saheri  
**Date:** May 2026  

The IBM HR Analytics dataset captures internal employee attributes — tenure, satisfaction, 
compensation, and demographics. Alone, it cannot answer whether attrition patterns at IBM 
reflect broader labour market conditions, or whether employee sentiment aligns with what 
the structured data suggests.

This notebook extends the analytical foundation across three additional data sources:

| Source | Type | Purpose |
|---|---|---|
| Synthetic Payroll | Simulated internal records | Month-level compensation trends across 24 months |
| BLS Public API | Official labour statistics | Macro context: quit rates, unemployment, wage growth |
| AmbitionBox | Scraped employee reviews | Qualitative sentiment: what employees actually say about IBM |

Each source introduces deliberate complexity — missing values, duplicates, unstructured text — 
reflecting the messy reality of real-world data pipelines.

## 1. Setup & Data Loading
Import all required libraries and load the raw IBM HR Analytics dataset from the 
`data/raw/` directory. No transformations are applied at this stage.

In [2]:
#Import necessary libraries
import os
import json
import time
import warnings
import requests
import numpy as np
import pandas as pd
import random
import re
import csv
from faker import Faker
from bs4 import BeautifulSoup
from dotenv import load_dotenv
from datetime import datetime

warnings.filterwarnings('ignore')
load_dotenv()

True

In [3]:
#Load the dataset   
BASE_DIR = os.path.dirname(os.path.abspath('__file__'))
data = pd.read_csv(os.path.join(BASE_DIR, '..', 'data', 'raw', 'WA_Fn-UseC_-HR-Employee-Attrition.csv'))

## 2. Synthetic Payroll Data

The IBM dataset provides a single `MonthlyIncome` snapshot per employee — a point-in-time figure with no history.
Real attrition analysis, however, often depends on compensation trajectory: whether salaries are growing, stagnating, or being frozen.

To simulate this, we generate 24 months of payroll records (2024–2025) for all 1,470 employees, using IBM's `MonthlyIncome` as the baseline salary for each individual.

### Design Decisions

**Salary Freeze (Q3 2024):** A firm-wide salary freeze is applied across July–September 2024, simulating a realistic cost-control event. Increments scheduled during this window are suppressed.

**Performance-Linked Bonuses:** Quarterly bonuses (March, June, September, December) scale with IBM's `PerformanceRating` column — high performers receive 10–22%, average performers 4–10%.

**Overtime:** Employees flagged `OverTime = Yes` in the IBM dataset have a 75% monthly probability of earning overtime; others have a 10% probability.

### Deliberate Messiness
To reflect real-world data quality issues, three types of noise are introduced:
- **Duplicate rows** — simulating data entry errors across 50 employees
- **Missing OvertimePay** — 5% of rows set to `NaN`
- **Outlier bonuses** — 20 rows inflated 10× to simulate keying errors

These will be addressed during the cleaning phase in Notebook 05.

In [4]:
#Initialize Faker and set seeds for reproducibility
fake = Faker()
Faker.seed(42)
random.seed(42)
np.random.seed(42)

In [5]:
# Verify required columns exist
required_columns = [
    "EmployeeNumber",
    "MonthlyIncome",
    "OverTime",
    "PerformanceRating"
]
missing = [col for col in required_columns if col not in data.columns]
if missing:
    raise ValueError(f"Missing columns in IBM dataset: {missing}")

In [6]:
# Ensure we have 1470 employees to generate 24 months of data for each employee (1470 * 24 = 35280 records)
data = data.head(1470).copy()

In [7]:
# Generate synthetic employee names using Faker
data["EmployeeName"] = [
    fake.name() for _ in range(len(data))
]

In [8]:
# Define years and months for the synthetic data generation
current_year = datetime.now().year
years = [current_year - 2, current_year - 1]
months = list(range(1, 13))

In [9]:
# Generate synthetic payroll data for 24 months for each employee
payroll_rows = []

for _, row in data.iterrows():

    employee_id = row["EmployeeNumber"]
    employee_name = row["EmployeeName"]
    original_salary = row["MonthlyIncome"]
    overtime_flag = row["OverTime"]
    performance = row["PerformanceRating"]

    freeze_quarters = [(2024, 7), (2024, 8), (2024, 9)]

    increment_month = random.randint(12, 18)
    increment_percent = random.uniform(0.05, 0.18)

    if performance == 4:
        bonus_percent = random.uniform(0.10, 0.22)
    elif performance == 3:
        bonus_percent = random.uniform(0.04, 0.10)
    else:
        bonus_percent = random.uniform(0.01, 0.03)

    month_counter = 0
    current_salary = original_salary

    for year in years:
        for month in months:

            month_counter += 1

            in_freeze = (year, month) in freeze_quarters

            if month_counter == increment_month and not in_freeze:
                current_salary = current_salary * (1 + increment_percent)

            # Overtime Logic
            if overtime_flag == "Yes":
                overtime_probability = 0.75
            else:
                overtime_probability = 0.10

            if random.random() < overtime_probability:
                overtime_pay = np.random.normal(
                    loc=current_salary * 0.08,
                    scale=current_salary * 0.03
                )
                overtime_pay = max(0, overtime_pay)
            else:
                overtime_pay = 0

            # Bonus Logic
            bonus_months = [3, 6, 9, 12]

            if month in bonus_months:
                if performance == 4:
                    bonus = current_salary * bonus_percent
                elif performance == 3:
                    if random.random() < 0.80:
                        bonus = current_salary * bonus_percent
                    else:
                        bonus = 0
                else:
                    bonus = current_salary * bonus_percent
            else:
                bonus = 0

            # Total compensation
            total_compensation = (
                current_salary
                + overtime_pay
                + bonus
            )

            payroll_rows.append({
                "EmployeeNumber": employee_id,
                "EmployeeName": employee_name,
                "Month": month,
                "Year": year,
                "BaseSalary": round(current_salary, 2),
                "OvertimePay": round(overtime_pay, 2),
                "Bonus": round(bonus, 2),
                "TotalCompensation": round(total_compensation, 2)
            })

In [10]:
# Create DataFrame from generated payroll data

payroll_df = pd.DataFrame(payroll_rows)

# Sort by EmployeeNumber, Year, Month for better readability

payroll_df = payroll_df.sort_values(
    by=["EmployeeNumber", "Year", "Month"]
)

In [11]:
# 1. Introduce duplicate rows (simulate data entry errors)
duplicate_employees = payroll_df['EmployeeNumber'].sample(50, random_state=42)
duplicates = payroll_df[payroll_df['EmployeeNumber'].isin(duplicate_employees)].sample(100, random_state=42)
payroll_df = pd.concat([payroll_df, duplicates]).reset_index(drop=True)

# 2. Introduce missing OvertimePay values randomly
missing_indices = payroll_df.sample(frac=0.05, random_state=42).index
payroll_df.loc[missing_indices, 'OvertimePay'] = np.nan

# 3. Introduce outlier bonuses (data entry errors)
outlier_indices = payroll_df.sample(20, random_state=42).index
payroll_df.loc[outlier_indices, 'Bonus'] = payroll_df.loc[outlier_indices, 'Bonus'] * 10

In [12]:
# Export to CSV
payroll_df.to_csv(
    os.path.join(BASE_DIR, '..', 'data', 'processed', 'synthetic_Payroll_24_Months.csv'),
    index=False
)

In [12]:
# Output summary of generated dataset
print("Synthetic Payroll Dataset Generated:")
print(f"Employees: {payroll_df['EmployeeNumber'].nunique()}")
print(f"Rows Generated: {len(payroll_df)}")

print("\nColumns:")
print(payroll_df.columns.tolist())

print("\nSample Data:")
print(payroll_df.head())

print("\nCSV SAVED:")
print("synthetic_payroll_24_months.csv")

Synthetic Payroll Dataset Generated:
Employees: 1470
Rows Generated: 35380

Columns:
['EmployeeNumber', 'EmployeeName', 'Month', 'Year', 'BaseSalary', 'OvertimePay', 'Bonus', 'TotalCompensation']

Sample Data:
   EmployeeNumber  EmployeeName  Month  Year  BaseSalary  OvertimePay   Bonus  \
0               1  Allison Hill      1  2024      5993.0       568.74    0.00   
1               1  Allison Hill      2  2024      5993.0       454.58    0.00   
2               1  Allison Hill      3  2024      5993.0       595.89  276.58   
3               1  Allison Hill      4  2024      5993.0       753.27    0.00   
4               1  Allison Hill      5  2024      5993.0          NaN    0.00   

   TotalCompensation  
0            6561.74  
1            6447.58  
2            6865.46  
3            6746.27  
4            6430.34  

CSV SAVED:
synthetic_payroll_24_months.csv


### Output
- **Rows generated:** 35,380 (35,280 base + 100 duplicates)
- **Employees covered:** 1,470
- **Period:** January 2024 – December 2025
- **Saved to:** `data/processed/synthetic_payroll_24_months.csv`

The dataset is intentionally imperfect. Missing values, duplicates, and outliers are features of this dataset by design — they exist to demonstrate data cleaning judgment in the next notebook.

## 3. Bureau of Labor Statistics (BLS) API

Internal compensation data alone cannot explain attrition timing. A useful question for TalentSight is: *were employees leaving IBM at a time when the external labour market was also hot?*

To answer this, we pull three macroeconomic indicators from the BLS Public API (v2) covering 2023–2025:

| Series ID | Metric | Relevance to Attrition |
|---|---|---|
| `JTS000000000000000QUR` | JOLTS Quit Rate | Measures voluntary separations across the US economy — a direct proxy for labour market confidence |
| `LNS14000000` | Unemployment Rate | Low unemployment = more outside options = higher attrition risk |
| `CIU1010000000000A` | Employment Cost Index | Tracks wage growth pressure — rising ECI signals competitors are paying more |

These series will be merged with internal attrition data during feature engineering to test whether macro conditions correlate with IBM-specific turnover patterns.

In [13]:
# Load environment variables
load_dotenv()
BLS_API_KEY = os.getenv("BLS_API_KEY")
print("API Key loaded:", BLS_API_KEY is not None)

API Key loaded: True


In [14]:
# Set up API request
headers = {'Content-type': 'application/json'}

payload = {
    "seriesid": [
        "JTS000000000000000QUR",
        "LNS14000000",
        "CIU1010000000000A"
    ],
    "startyear": "2023",
    "endyear": "2025",
    "calculations": True,
    "annualaverage": False,
    "registrationkey": BLS_API_KEY
}

response = requests.post(
    "https://api.bls.gov/publicAPI/v2/timeseries/data/",
    data=json.dumps(payload),
    headers=headers
)

print(response.status_code)
print(response.json()['status'])

200
REQUEST_SUCCEEDED


In [15]:
# Parse BLS API response into a dataframe
bls_data = response.json()['Results']['series']

bls_rows = []

for series in bls_data:
    series_id = series['seriesID']
    for record in series['data']:
        bls_rows.append({
            'SeriesID': series_id,
            'Year': record['year'],
            'Period': record['period'],
            'Value': record['value'],
            'PeriodName': record['periodName']
        })

bls_df = pd.DataFrame(bls_rows)
print(bls_df.shape)
print(bls_df.head(10))

(84, 5)
                SeriesID  Year Period Value PeriodName
0  JTS000000000000000QUR  2025    M12   2.0   December
1  JTS000000000000000QUR  2025    M11   2.0   November
2  JTS000000000000000QUR  2025    M10   1.9    October
3  JTS000000000000000QUR  2025    M09   1.9  September
4  JTS000000000000000QUR  2025    M08   2.0     August
5  JTS000000000000000QUR  2025    M07   2.0       July
6  JTS000000000000000QUR  2025    M06   2.1       June
7  JTS000000000000000QUR  2025    M05   2.1        May
8  JTS000000000000000QUR  2025    M04   2.0      April
9  JTS000000000000000QUR  2025    M03   2.2      March


In [16]:
# Save BLS data to raw folder
bls_df.to_csv(
    os.path.join(BASE_DIR, '..', 'data', 'raw', 'bls_labor_data.csv'),
    index=False
)
print("BLS data saved successfully")
print(f"Shape: {bls_df.shape}")

BLS data saved successfully
Shape: (84, 5)


### Output
- **Records retrieved:** 84 rows across 5 series
- **Period covered:** January 2023 – December 2025
- **Saved to:** `data/raw/bls_labor_data.csv`

All three series returned complete monthly data with no gaps. The JOLTS quit rate shows a declining trend from 2023 into 2025, consistent with a cooling labour market — relevant context for interpreting IBM's attrition pattern over the same window.

## 4. Glassdoor — Attempted Scrape (Blocked)

Glassdoor was the first-choice source for employee sentiment data, given its brand recognition and review volume for IBM.

A direct HTTP request with browser-mimicking headers was attempted against the Glassdoor reviews index.

In [17]:
#Scrape Glassdoor for reviews and ratings of IBM
url = "https://www.glassdoor.co.in/Reviews/index.htm"

headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/137.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "en-US,en;q=0.9"
}

response = requests.get(url, headers=headers)

print(response.status_code)

soup = BeautifulSoup(response.text, "html.parser")

print(soup.title)
print(response.text[:500])

403
<title>Security | Glassdoor</title>
<!doctype html>
<html lang="en">
  <head>
    <title>Security | Glassdoor</title>
    <meta http-equiv="content-type" content="text/html;charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="theme-color" content="#0caa41" />
    <meta name="robots" content="noindex, nofollow" />
    <link rel="icon" href="https://www.glassdoor.com/favicon.ico" sizes="any" />
    <style type="text/css">
      @font-face {
        font-family: "Glassdoor Sans"


### Result: 403 — Access Denied

Glassdoor returned a Cloudflare security challenge (HTTP 403) immediately on the first request, with no review content accessible.

This is expected behaviour. Glassdoor actively blocks programmatic access and requires either:
- A paid API partnership, or
- A JavaScript-rendered browser session (Selenium/Playwright)

Neither is appropriate for this project scope. **AmbitionBox** was selected as the fallback — it carries 25,000+ IBM reviews, covers the same sentiment dimensions, and its static HTML is accessible via standard HTTP requests.

> **Note for portfolio context:** Documenting a blocked scrape and pivoting with justification is more representative of real-world data collection than a clean scrape. Real data pipelines fail; what matters is how you respond.

## 4. AmbitionBox — Employee Sentiment Scraping

Structured HR data explains *what* happened. It cannot explain *why* employees left in their own words. To capture qualitative sentiment, we scrape IBM employee reviews from AmbitionBox — a platform hosting 25,000+ IBM reviews with sub-ratings across seven dimensions.

**Glassdoor** was the first target. A direct HTTP request returned a 403 Security block — Glassdoor employs bot detection that cannot be bypassed with header spoofing alone. AmbitionBox serves static HTML and is accessible via `requests` + `BeautifulSoup`.

Each review captures:

| Field | Description |
|---|---|
| `Overall Rating` | Reviewer's headline score out of 5 |
| `Work-Life Balance` | Sub-rating from the carousel |
| `Salary Rating` | Sub-rating from the carousel |
| `Culture Rating` | Sub-rating from the carousel |
| `Job Security Rating` | Sub-rating from the carousel |
| `Promotions Rating` | Sub-rating from the carousel |
| `Pros` | Free-text Likes section |
| `Cons` | Free-text Dislikes section |
| `Job Title` | Reviewer's role at IBM |
| `Employment Status` | Permanent, Former, Contract |
| `Review Date` | Date review was posted |

200 pages are scraped (~4,000 reviews) — sufficient for segment-level sentiment analysis without exhausting the source. Sub-ratings are extracted by matching label text rather than positional index, since not every reviewer fills all seven dimensions. Pros and Cons are matched via their `h3` header siblings, handling reviews where only one section is present.

These reviews will be used in feature engineering to construct sentiment scores that complement the structured IBM attrition signals.

In [18]:
#Scrape AmbitionBox for reviews and ratings of IBM
url = "https://www.ambitionbox.com/reviews/ibm-reviews"

headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/137.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "en-US,en;q=0.9"
}

response = requests.get(url, headers=headers)

print(response.status_code)

soup = BeautifulSoup(response.text, "html.parser")

print(soup.title)
print(response.text[:500])

200
<title>IBM Reviews by 25000+ Employees | Rated 3.9/5 | AmbitionBox</title>
<!DOCTYPE html><html><head><meta charSet="utf-8"/><title>IBM Reviews by 25000+ Employees | Rated 3.9/5 | AmbitionBox</title><meta name="title" content="IBM Reviews by 25000+ Employees | Rated 3.9/5 | AmbitionBox"/><meta name="description" content="IBM is rated 3.9/5, based on 25000+ employee reviews on AmbitionBox. IBM is highly rated for Work-life balance with 4.0/5."/><meta name="viewport" content="width=device-width,initial-scale=1, minimum-scale=1, interactive-widget=resizes-content"/><link 


In [ ]:
# Try to find review blocks on AmbitionBox using the class name from the HTML
response = requests.get(
    "https://www.ambitionbox.com/reviews/ibm-reviews?page=1",
    headers=headers
)
soup = BeautifulSoup(response.text, "html.parser")

reviews = soup.find_all(
    "div",
    class_="font-pn-400 pt-0 px-4 pb-1 -mx-4 md:-mx-5 relative"
)

if reviews:
    print(f"Found {len(reviews)} review blocks")
    print("\n--- RAW HTML OF FIRST REVIEW ---")
    print(reviews[0].prettify())
else:
    print("No review blocks found — outer selector is broken too")

Found 20 review blocks

--- RAW HTML OF FIRST REVIEW ---
<div aria-hidden="true" class="font-pn-400 pt-0 px-4 pb-1 -mx-4 md:-mx-5 relative" id="review-79915127" style="content-visibility:auto;contain-intrinsic-size:auto 1000px">
 <div class="-mx-4 px-3 flex justify-between border-l-4 border-rating-1.5 -mx-4 mt-4" data-testid="ReviewCard_79915127_Header" testid="ReviewCard_79915127_Header">
  <div class="flex-grow flex flex-col items-start md:items-baseline relative max-w-full">
   <div class="w-full">
    <div class="w-full">
     <div class="w-full flex justify-between" data-testid="ReviewCard_79915127_RatingRow" testid="ReviewCard_79915127_RatingRow">
      <div class="gap-1 flex flex-row items-center" data-testid="">
       <svg class="" fill="none" height="20" viewbox="0 0 24 24" width="20" xmlns="http://www.w3.org/2000/svg">
        <rect fill="#ff6e2b" height="24" rx="4.8" width="24">
        </rect>
        <path d="M7.36875 19.5L8.5875 13.9539L4.5 10.2237L9.9 9.73026L12 4.5L14.

In [ ]:
# Loop through multiple pages of reviews, extract relevant data, and build a dataframe

all_reviews = []

for page in range(1, 200):

    print(f"Scraping page {page}...")

    url = f"https://www.ambitionbox.com/reviews/ibm-reviews?page={page}"

    try:
        response = requests.get(url, headers=headers, timeout=15)
    except requests.exceptions.RequestException as e:
        print(f"Request failed on page {page}: {e}")
        break

    if response.status_code != 200:
        print(f"Blocked on page {page} — status {response.status_code}. Stopping.")
        break

    soup = BeautifulSoup(response.text, "html.parser")

    reviews = soup.find_all(
        "div",
        class_="font-pn-400 pt-0 px-4 pb-1 -mx-4 md:-mx-5 relative"
    )

    if not reviews:
        print(f"No reviews on page {page}. Scraping complete.")
        break

    print(f"  Found {len(reviews)} reviews")

    for i in reviews:

        # Overall Rating
        rating = None
        rating_tag = i.find(
            "span",
            class_="text-primary-text font-pn-700 text-xl leading-none"
        )
        if rating_tag:
            rating = rating_tag.get_text(strip=True)

        # Review Date + Job Title
        # Both live inside the h2 block
        review_date = None
        job_title = None

        h2_tag = i.find("h2", class_="m-0 leading-4 inline")
        if h2_tag:
            h2_text = h2_tag.get_text(" ", strip=True)
            match = re.search(r"rated by a[n]?\s+(.*?)\s+in", h2_text)
            if match:
                job_title = match.group(1)

        date_span = i.find("span", class_="leading-4")
        if date_span:
            raw = date_span.get_text(" ", strip=True)
            review_date = raw.replace("on", "").replace('"', "").strip()

        # Ratings Carousel (WLB + others)
        # Each sub-rating is a div with class "flex flex-row items-center gap-1.5 flex-shrink-0"
        # Structure: img | score span | label span
        wlb = None
        salary_rating = None
        culture_rating = None
        job_security_rating = None
        promotions_rating = None
        skill_dev_rating = None
        work_satisfaction_rating = None

        carousel_items = i.find_all(
            "div",
            class_="flex flex-row items-center gap-1.5 flex-shrink-0"
        )

        for item in carousel_items:
            spans = item.find_all("span")
            if len(spans) >= 2:
                score = spans[0].get_text(strip=True)
                label = spans[1].get_text(strip=True).lower()

                if "work-life" in label or "work life" in label:
                    wlb = score
                elif "salary" in label:
                    salary_rating = score
                elif "culture" in label:
                    culture_rating = score
                elif "job security" in label:
                    job_security_rating = score
                elif "promotion" in label:
                    promotions_rating = score
                elif "skill" in label:
                    skill_dev_rating = score
                elif "satisfaction" in label:
                    work_satisfaction_rating = score

        #  Pros and Cons 
        # Match h3 header ("Likes" / "Dislikes") then take the following p sibling
        pros = None
        cons = None

        body_div = i.find("div", attrs={"data-testid": lambda x: x and "Body" in x})

        if body_div:
            h3_tags = body_div.find_all("h3")
            for h3 in h3_tags:
                header_text = h3.get_text(strip=True).lower()
                # Find the next p sibling
                next_p = h3.find_next_sibling("p")
                if next_p:
                    text = next_p.get_text(strip=True)
                    if "like" in header_text:   # "Likes"
                        pros = text
                    elif "dislike" in header_text:  # "Dislikes"
                        cons = text

        # Employment Status
        employment_status = None
        status_span = i.find("span", class_="flex-shrink-0 flex items-center")
        if status_span:
            parts = status_span.get_text(" ", strip=True).split("·")
            if len(parts) > 1:
                employment_status = parts[1].strip()

        all_reviews.append({
            "Overall Rating": rating,
            "Review Date": review_date,
            "Job Title": job_title,
            "Work-Life Balance": wlb,
            "Salary Rating": salary_rating,
            "Culture Rating": culture_rating,
            "Job Security Rating": job_security_rating,
            "Promotions Rating": promotions_rating,
            "Skill Development Rating": skill_dev_rating,
            "Work Satisfaction Rating": work_satisfaction_rating,
            "Pros": pros,
            "Cons": cons,
            "Employment Status": employment_status
        })

    time.sleep(2)


#  Build and save 
IBM_review = pd.DataFrame(all_reviews)
IBM_review.drop_duplicates(inplace=True)
IBM_review.reset_index(drop=True, inplace=True)

IBM_review.to_csv(
    os.path.join(BASE_DIR, '..', 'data', 'scraped', 'ibm_reviews.csv'),
    index=False,
    quoting=csv.QUOTE_ALL,
    lineterminator='\n'
)

print(f"\nScraping complete — {len(IBM_review)} reviews saved")
print(f"WLB coverage   : {IBM_review['Work-Life Balance'].notna().sum()} / {len(IBM_review)}")
print(f"Pros coverage  : {IBM_review['Pros'].notna().sum()} / {len(IBM_review)}")
print(f"Cons coverage  : {IBM_review['Cons'].notna().sum()} / {len(IBM_review)}")
print(IBM_review.head())


Scraping page 1...
  Found 20 reviews
Scraping page 2...
  Found 20 reviews
Scraping page 3...
  Found 20 reviews
Scraping page 4...
  Found 20 reviews
Scraping page 5...
  Found 20 reviews
Scraping page 6...
  Found 20 reviews
Scraping page 7...
  Found 20 reviews
Scraping page 8...
  Found 20 reviews
Scraping page 9...
  Found 20 reviews
Scraping page 10...
  Found 20 reviews
Scraping page 11...
  Found 20 reviews
Scraping page 12...
  Found 20 reviews
Scraping page 13...
  Found 20 reviews
Scraping page 14...
  Found 20 reviews
Scraping page 15...
  Found 20 reviews
Scraping page 16...
  Found 20 reviews
Scraping page 17...
  Found 20 reviews
Scraping page 18...
  Found 20 reviews
Scraping page 19...
  Found 20 reviews
Scraping page 20...
  Found 20 reviews
Scraping page 21...
  Found 20 reviews
Scraping page 22...
  Found 20 reviews
Scraping page 23...
  Found 20 reviews
Scraping page 24...
  Found 20 reviews
Scraping page 25...
  Found 20 reviews
Scraping page 26...
  Found 20 rev

### Output
- **Pages scraped:** 200
- **Total reviews collected:** *(fill from output)*
- **Reviews after deduplication:** *(fill from output)*
- **Saved to:** `data/scraped/ibm_reviews.csv`

Work-Life Balance and sub-rating coverage confirms the carousel selector is parsing correctly. Pros and Cons fields are saved with full CSV quoting — both sections contain natural newlines that would otherwise corrupt row boundaries. Some reviews contain only Dislikes with no Likes section; null Pros values on those rows are valid data, not missing data.

---
### 5. Data Collection Summary

| Dataset | Rows | Destination | Status |
|---|---|---|---|
| Synthetic Payroll | 35,380 | `data/processed/` | ✅ Complete |
| BLS Labour Data | 84 | `data/raw/` | ✅ Complete |
| AmbitionBox Reviews | ~2,000 | `data/scraped/` | ✅ Complete |
| Glassdoor Reviews | — | — | ❌ Blocked (403) |

The project now holds structured internal data (IBM HR + Payroll), macroeconomic 
context (BLS), and unstructured employee sentiment (AmbitionBox). 

The next notebook — **05_Feature_Engineering** — will merge these sources, resolve 
the data quality issues introduced in the payroll data, and prepare a modelling-ready 
feature set.